### PR TITLE
support mpp_prober to detect tiflash aliveness

### PR DESCRIPTION
**Pull request by guo-shaoge**
Basically same as https://github.com/pingcap/tidb/blob/master/store/copr/mpp_probe.go

1. MPPProber record all failed stores, and will check if alive in background.
2. Caller first check if in MPPProber.failed_stores, if true, then ignore this failed store, otherwise send alive rpc to do real detect.
3. For now, thread pool only has one thread. Will add another background thread to refresh region cache later.
4. test case will be add in endless later, only manually test for now.

### CODE REVIEW

File: include/pingcap/kv/Cluster.h [hunk 1]
```diff
@@ -27,10 +29,17 @@ struct Cluster
...
     ::kvrpcpb::APIVersion api_version = ::kvrpcpb::APIVersion::V1;
 
+    std::unique_ptr<pingcap::common::FixedThreadPool> thread_pool;
+    std::unique_ptr<common::MPPProber> mpp_prober;
+
     Cluster()
         : pd_client(std::make_shared<pd::MockPDClient>())
         , rpc_client(std::make_unique<RpcClient>())
-    {}
+        , thread_pool(std::make_unique<pingcap::common::FixedThreadPool>(1))
```
Comment (guo-shaoge) [inline include/pingcap/kv/Cluster.h:38]: Fow now, one thread is enough. We may refresh region cache in background later, at that time, we may need multiple threads.

File: include/pingcap/common/MPPProber.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    Logger * log;
+    TimePoint recovery_timepoint;
+    TimePoint last_lookup_timepoint;
+    TimePoint last_detect_timepoint;
+    std::mutex state_lock;
+
+    void detectAndUpdateState(const std::chrono::seconds & detect_period, size_t detect_rpc_timeout);
+};
+
+class MPPProber
```
Comment (JaySon-Huang) [inline include/pingcap/common/MPPProber.h:59]: Better add some description of the functionality of this struct

File: src/coprocessor/Client.cc [hunk 3]
```diff
@@ -115,7 +115,23 @@ std::vector<LocationKeyRanges> splitKeyRangesByLocations(
...
+    std::map<uint64_t, kv::Store> alive_stores;
+    for (const auto & ele : stores)
+    {
+        if (!cluster->mpp_prober->isRecovery(ele.second.addr, /*mpp_fail_ttl=*/std::chrono::seconds(60)))
+            continue;
+
+        if (!common::detectStore(cluster->rpc_client, ele.second.addr, /*rpc_timeout=*/2, log))
+            cluster->mpp_prober->add(ele.second.addr);
+
+        alive_stores.emplace(ele.first, ele.second);
```
Comment (JaySon-Huang) [inline src/coprocessor/Client.cc:129]: why detectStore return not available, still needs to add the store to `alive_stores`?

File: src/common/MPPProber.cc [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        for (const auto & store : recovery_stores)
+        {
+            failed_stores.erase(store);
+        }
+    }
+}
+
+void ProbeState::detectAndUpdateState(const std::chrono::seconds & detect_period, size_t detect_rpc_timeout)
+{
+    if (getElapsed(last_detect_timepoint) < detect_period)
```
Comment (JaySon-Huang) [inline src/common/MPPProber.cc:104]: where do we update this value?

File: src/common/MPPProber.cc [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        std::lock_guard<std::mutex> lock(store_lock);
+        copy_failed_stores = failed_stores;
+    }
+    auto iter = copy_failed_stores.find(store_addr);
+    if (iter == copy_failed_stores.end())
+        return true;
+
+    {
+        std::lock_guard<std::mutex> lock(iter->second->state_lock);
+        return iter->second->recovery_timepoint != INVALID_TIME_POINT && getElapsed(iter->second->recovery_timepoint) > recovery_ttl;
```
Comment (JaySon-Huang) [inline src/common/MPPProber.cc:26]: why here need `getElapsed(iter->second->recovery_timepoint) > recovery_ttl`?

Comment (guo-shaoge) [reply src/common/MPPProber.cc:26]: Max recovery time is `15mins`, recovery_ttl is 60s. See logic in tidb code.

File: src/common/MPPProber.cc [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    }
+}
+
+void MPPProber::add(const std::string & store_addr)
+{
+    std::lock_guard<std::mutex> lock(store_lock);
+    auto iter = failed_stores.find(store_addr);
+    if (iter == failed_stores.end())
+    {
+        failed_stores[store_addr] = std::make_shared<ProbeState>(store_addr, cluster);
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply src/common/MPPProber.cc:36]: Yes

File: src/common/MPPProber.cc (hunk 4 above)
Comment (guo-shaoge) [reply src/common/MPPProber.cc:104]: done

File: src/common/MPPProber.cc (hunk 5 above)
Comment (guo-shaoge) [reply src/common/MPPProber.cc:27]: Aka, request will succeed after `60secs` if send to this recoveried store explicitly.
background scan will remove this failed store after `15mins`.

**Review by JaySon-Huang — APPROVED** [review]

**Review by gengliqi — APPROVED** [review]

**Pull request merged**